Improve invite users to organization feature

klakhov opened the pull request:
<!-- Raise an issue to propose your change (https://github.com/opencv/cvat/issues).
It helps to avoid duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [Contribution guide](https://opencv.github.io/cvat/docs/contributing/). -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

This PR improves workflow with invitations to organizations. It allows to view, accept, decline invitations to organizations (previously all the invitations were auto-accepted).
It fixes problems with inviting unregistered users to organization allowing them to register without invitation. 

![image](https://github.com/opencv/cvat/assets/50956430/71bededa-be48-4e7a-80b6-febba86c2a3f)
![image](https://github.com/opencv/cvat/assets/50956430/773119a2-4d88-46c1-bad6-ed24734ce0c2)


### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->
The best way to test this PR is to add those lines to `cvat/settings/base.py`:
Enable email verification:
```
ACCOUNT_AUTHENTICATION_METHOD = 'username_email'
ACCOUNT_CONFIRM_EMAIL_ON_GET = True
ACCOUNT_EMAIL_REQUIRED = True
ACCOUNT_EMAIL_VERIFICATION = 'mandatory'
```

Setup console email backend
```
# Email backend settings for Django
EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
```

And edit `ORG_INVITATION_CONFIRM` to `Yes`
```
ORG_INVITATION_CONFIRM = 'Yes'
```
### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable for some reason, then ~~explicitly strikethrough~~ the whole
line. If you don't do that, GitHub will show incorrect progress for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment <!-- see top comment in CHANGELOG.md -->
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

bsekachev commented:
It looks like being an organization owner I can myself accept any invitation knowing its key and email (key may be accessed in GET request when open organizations page) 

<img width="1076" alt="image" src="https://github.com/opencv/cvat/assets/40690378/5f32e717-eb8e-4300-a644-d45858a70df1">

I may just open the link and press Ok, after that invitation will be accepted. Looks like potential vulnerability

<img width="672" alt="image" src="https://github.com/opencv/cvat/assets/40690378/616577d8-a00f-48e0-83f1-2e7b5410ee84">

bsekachev commented:
1. I do not think we need to reload the page when click "Cancel" button.
2. I believe we need "Accept", "Reject" instead of "Ok", "Cancel"
![issue_1](https://github.com/opencv/cvat/assets/40690378/459ffcc7-bea6-4c81-ba18-a7f8e7aedf9e)

bsekachev commented:
If invited user is already in another organization, I can't get invitation (404 code) 
![issue_2](https://github.com/opencv/cvat/assets/40690378/bd6b5c29-4dbd-40d9-a37c-39951226b12c)

bsekachev commented:
I believe we should not show organization where the user was invited in a list of organizations until invitation is accepted
<img width="278" alt="image" src="https://github.com/opencv/cvat/assets/40690378/2dd8713b-1af0-4ced-b8ce-2852f52338e6">

<img width="297" alt="image" src="https://github.com/opencv/cvat/assets/40690378/ccfdbcc4-f44e-4e09-930d-ff9f3c64452f">

bsekachev commented:
We should make the field email readonly if there is active invitation:

<img width="1110" alt="image" src="https://github.com/opencv/cvat/assets/40690378/f8d34eec-0a55-422a-afc5-c291a0e79ae8">

Otherwise I see strange behaviour, even if I changed email:

<img width="453" alt="image" src="https://github.com/opencv/cvat/assets/40690378/ad295e50-1818-4b8d-929a-24ea2685af8e">

It anyway sends confirmation letter to email from invitation, the field is reset after this.

<img width="622" alt="image" src="https://github.com/opencv/cvat/assets/40690378/5d83ca00-4942-4b2c-916e-8741e530145b">

bsekachev commented on cvat-ui/src/components/invitation-watcher/invitation-watcher.tsx:
```diff
@@ -0,0 +1,102 @@
+                invitationsToBeShown.forEach((invitation: Invitation, index) => {
+                    const slug = invitation.organization instanceof Organization ? invitation.organization.slug : '';
+                    Modal.confirm({
+                        title: `You've been invited to an organization ${slug} by ${invitation.owner?.username}`,
```
```suggestion
                        title: `You've been invited to the organization ${slug} by ${invitation.owner?.username}`,
```

Let's make organization name and inviting username bold

bsekachev commented on cvat-ui/src/components/invitation-watcher/invitation-watcher.tsx:
```diff
@@ -0,0 +1,102 @@
+                    const slug = invitation.organization instanceof Organization ? invitation.organization.slug : '';
+                    Modal.confirm({
+                        title: `You've been invited to an organization ${slug} by ${invitation.owner?.username}`,
+                        content: 'Do you want to join?',
```
```suggestion
                        content: 'Would you like to join?',
```

bsekachev commented on cvat-ui/src/components/invitation-watcher/invitation-watcher.tsx:
```diff
@@ -0,0 +1,102 @@
+                        className: 'cvat-invitaion-confirm-modal',
+                        onOk: () => {
+                            dispatch(acceptInvitationAsync(invitation.key, (orgSlug) => {
+                                localStorage.setItem('currentOrganization', orgSlug);
+                                if (index === invitationsToBeShown.length - 1) {
+                                    window.location.reload();
+                                    localStorage.removeItem('invitations');
+                                }
```
As I remember dispatch returns promise in case with async function, so, we do not need callbacks here, just dispatch acceptInvitationAsync(invitation.key).then((orgSlug) => ...)

bsekachev commented on cvat-core/src/api-implementation.ts:
```diff
@@ -346,27 +346,34 @@ export default function implementAPI(cvat) {
         return orgSlug;
     };
 
+    cvat.organizations.rejectInvitation.implementation = async (
+        key,
+    ) => {
+        await serverProxy.organizations.rejectInvitation(
+            key,
+        );
+    };
```
```suggestion
    cvat.organizations.rejectInvitation.implementation = serverProxy.organizations.rejectInvitation;
```

bsekachev commented on cvat-ui/src/components/invitation-watcher/invitation-watcher.tsx:
```diff
@@ -0,0 +1,102 @@
+                    notification.error({
+                        message: 'Invitation error',
+                        className: 'cvat-invitation-error',
+                        description: `The invitation to ${invitation.organization.slug} is expired. ` +
```
```suggestion
                        description: `The invitation to ${invitation.organization.slug} has expired. ` +
```

bsekachev commented on cvat-ui/src/components/invitation-watcher/invitation-watcher.tsx:
```diff
@@ -0,0 +1,102 @@
+                const expiredInvitations = validInvitations.filter((invitation) => invitation.expired);
+                expiredInvitations.forEach((invitation) => {
+                    notification.error({
+                        message: 'Invitation error',
+                        className: 'cvat-invitation-error',
```
Expired invitation is not an error, it should be informational message

bsekachev commented on cvat-ui/src/components/invitation-watcher/invitation-watcher.tsx:
```diff
@@ -0,0 +1,102 @@
+                }
+                invitationsToBeShown.forEach((invitation: Invitation, index) => {
+                    const slug = invitation.organization instanceof Organization ? invitation.organization.slug : '';
+                    Modal.confirm({
```
With forEach it will just show all the modals (I believe this is already not a good UX) at the same time, the latest with highest index probably, and after submitting the latest, it will remove all the other invitations from local storage and will update the page immediately

bsekachev commented on cvat-ui/src/components/invitation-watcher/invitation-watcher.tsx:
```diff
@@ -0,0 +1,102 @@
+
+                promises.forEach((promise) => {
+                    if (promise.status === 'rejected') {
+                        notification.error({
```
Shall we remove not valid invitations?

klakhov replied:
Now on invitations page we show list on invitaions. Valid and epired ones

bsekachev commented on cvat-ui/src/utils/hooks.ts:
```diff
@@ -122,3 +123,27 @@ export function useCardHeightHOC(params: ICardHeightHOC): () => string {
     };
 }
+
+export function authQuery(queryParams: URLSearchParams): Record<string, string> | null {
```
This is not a hook. I am not sure this file is a good place for this function

klakhov replied:
Moved to utils

bsekachev commented on cvat-ui/src/actions/auth-actions.ts:
```diff
@@ -244,3 +234,19 @@ export const acceptInvitationAsync = (
+    dispatch(authActions.rejectInvitation());
+
+    try {
+        await cvat.organizations.rejectInvitation(
+            key,
+        );
```
```suggestion
        await cvat.organizations.rejectInvitation(key);
```

klakhov commented on cvat/apps/organizations/models.py:
```diff
@@ -66,11 +66,13 @@ def organization_id(self):
+                days=settings.ORG_INVITATION_EXPIRY_DAYS,
+            )
+            return expiration_date <= timezone.now()
+        return True
```
Fixed to `None`

klakhov commented on cvat/apps/organizations/views.py:
```diff
@@ -281,3 +284,13 @@ def resend(self, request, pk):
+    @action(detail=True, methods=['POST'], url_path='reject')
+    def reject(self, request, pk):
+        try:
+            invitation = Invitation.objects.get(key=pk)
```
Added OPA rules for `reject` endpoint

klakhov commented on cvat/apps/organizations/views.py:
```diff
@@ -281,3 +284,13 @@ def resend(self, request, pk):
+        try:
+            invitation = Invitation.objects.get(key=pk)
+            membership = invitation.membership
+            membership.delete()
```
If we just delete invitation, the membership still exists. But invitation is deleted  automatically if we delete membership

klakhov commented on cvat/apps/organizations/serializers.py:
```diff
@@ -62,7 +60,7 @@ class InvitationReadSerializer(serializers.ModelSerializer):
     class Meta:
         model = Invitation
-        fields = ['key', 'created_date', 'owner', 'role', 'user', 'organization']
+        fields = ['key', 'created_date', 'owner', 'role', 'user', 'organization', 'expired']
```
I think its good for user to see that there was an invitation but its expired. UI can show expired invitations. Otherwise user can think that there was an error and invitaion was never created.

klakhov commented:
@bsekachev 

- Fixed accepting/rejecting invitations from other users
- Removed reload on 'Reject'
- Changed ok/cancel to accept/reject
- Fixed if invited user is already in organization he cant accept/reject new invitaion
- Fixed list of organizations in top menu, now only active memberships organizations are shown
- Made email on register form readonly if email from invitation is specified

And for the latter comment:
- Added `/invitaions` page instead of multiple modals, with a list of invitations.

klakhov commented on cvat-core/src/api-implementation.ts:
```diff
@@ -346,27 +346,34 @@ export default function implementAPI(cvat) {
+    cvat.organizations.invitation.implementation = async (key) => {
+        const invitation = await serverProxy.organizations.invitation(key);
+
+        // When we request invitation by key we need to load organization
```
Added this information to Invitation from server

bsekachev commented:
Just noticed that invitation confirmation emails in debug env goes to **example.com** domain, at the same time registration confirmation goes to correct domain **localhost:3000**.

bsekachev commented:
When clicking Reject, modal text looks weird. I believe it should be simpler, like: 'Would you like to reject the invitation to the "OrgName" organization?'. Another solution is just do not show this additional modal, I do not really think we need it. Button Reject on the page should be `dangerous`. 
![image](https://github.com/opencv/cvat/assets/40690378/45c3ddaa-3aa5-4ea2-94d1-f3d45cac6daa)

nmanovic commented on cvat/apps/iam/permissions.py:
```diff
@@ -95,10 +95,12 @@ def get_iam_context(request, obj):
-        raise PermissionDenied({'message': 'You should be an active member in the organization'})
+        # Allow inactive members view, accept, reject their invitations
+        view = request.parser_context.get('view')
+        if not (view.basename == 'invitation' and view.action in ['retrieve', 'accept', 'reject']):
```
Why do we need the logic here?

klakhov replied:
PermissionDenied error is raised here if organization is not None and membership is None. That is the case for Invitation objects(they belong to some org and user that tries to accept invitation is not an active member). So we handle this case and do not throw an error.

nmanovic replied:
From the first look it is wrong place for the logic.

bsekachev commented on cvat-core/src/api-implementation.ts:
```diff
@@ -345,28 +345,24 @@ export default function implementAPI(cvat) {
-        userConfirmations,
-        key,
-    ) => {
+    cvat.organizations.acceptInvitation.implementation = async (key) => {
```
Now the implementation here looks useless.
Will it work? (The same for rejectInvitation)

``cvat.organizations.acceptInvitation.implementation = serverProxy.organizations.acceptInvitation; ``

klakhov replied:
It doesnt really work with Jest tests. If we dont use such implementation(even dummy). The error is thrown:
```
TypeError: Cannot read properties of undefined (reading 'acceptInvitation')
cvat.organizations.acceptInvitation.implementation = serverProxy.organizations.acceptInvitation;
```

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -2053,12 +2058,28 @@ async function deleteOrganizationMembership(membershipId: number): Promise<void>
+
+        response = await Axios.get(`${backendAPI}/invitations`, {
+            params: {
+                ...filter,
```
Do we need InvitationsFilter type? It has only one field ``key`` that is used above. Here is always empty object.

klakhov replied:
Removed that.

bsekachev commented:
For expired invitation:
![image](https://github.com/opencv/cvat/assets/40690378/08058e78-3261-48aa-a035-50b480805027)

Reject -> Remove (without any Modal confirmations)

bsekachev commented on cvat-ui/src/components/invitations-page/invitations-list.tsx:
```diff
@@ -0,0 +1,98 @@
+    query: InvitationsQuery;
+}
+
+const PAGE_SIZE = 9;
```
Why 9? Is the page size for server request the same? I could not find explicitly defined page size for such requests in client code

klakhov replied:
Answered about pagination in the comment below

bsekachev commented on cvat-ui/src/reducers/notifications-reducer.ts:
```diff
@@ -419,6 +469,51 @@ export default function (state = defaultState, action: AnyAction): Notifications
+                        rejectingInvitation: {
+                            message: 'Could not reject invitation',
+                            reason: action.payload.error,
+                            shouldLog: !(action.payload.error instanceof ServerError),
```
Probably class name is missed

bsekachev commented on cvat-ui/src/reducers/notifications-reducer.ts:
```diff
@@ -419,6 +469,51 @@ export default function (state = defaultState, action: AnyAction): Notifications
+                    ...state.messages,
+                    invitations: {
+                        ...state.messages.invitations,
+                        resendingInvitation: 'Invintation was sent sucessfully',
```
```suggestion
                        resendingInvitation: 'Invitation was sent successfully',
```

bsekachev commented on cvat-ui/src/reducers/notifications-reducer.ts:
```diff
@@ -419,6 +469,51 @@ export default function (state = defaultState, action: AnyAction): Notifications
+                            message: 'Could not resend invitation',
+                            reason: action.payload.error,
+                            shouldLog: !(action.payload.error instanceof ServerError),
+                            className: 'cvat-notification-notice-resend-organization-invintation-failed',
```
```suggestion
                            className: 'cvat-notification-notice-resend-organization-invitation-failed',
```

bsekachev commented on cvat-ui/src/reducers/notifications-reducer.ts:
```diff
@@ -390,26 +397,69 @@ export default function (state = defaultState, action: AnyAction): Notifications
+                        fetching: {
+                            message: 'Could not get invitations',
+                            reason: action.payload.error,
+                            shouldLog: !(action.payload.error instanceof ServerError),
```
Probably class name is missed

bsekachev commented on cvat-ui/src/reducers/notifications-reducer.ts:
```diff
@@ -390,26 +397,69 @@ export default function (state = defaultState, action: AnyAction): Notifications
+            }
+            return state;
+        }
+        case InvitationsActionTypes.ACCEPT_INVITATION_SUCCESS: {
+            return {
+                ...state,
+                messages: {
+                    ...state.messages,
+                    invitations: {
+                        ...state.messages.invitations,
+                        acceptInvitationDone: 'Invitation accepted successfully.',
+                    },
+                },
+            };
+        }
+        case InvitationsActionTypes.REJECT_INVITATION_SUCCESS: {
```
I do not really think we need notifications when invitations accepted/rejected. 
It is quite natural that it was successful if errors not appeared (if only it is not long process, like exporting a dataset)
Of course we need some kind UI indication, but probably we already have on invitations page. 

Up to you, just my opinion here

klakhov replied:
Removed them.

bsekachev commented on cvat-ui/src/reducers/notifications-reducer.ts:
```diff
@@ -390,26 +397,69 @@ export default function (state = defaultState, action: AnyAction): Notifications
+                        ...state.messages,
+                        invitations: {
+                            ...state.messages.invitations,
+                            newInvitations: 'You have new organization invitations. [Click here](/invitations) to view them.',
```
I would suggest something simpler: 

Somebody invited you to an organization. [Click here](/invitations) to get details

bsekachev commented on cvat-ui/src/components/invitations-page/invitations-list.tsx:
```diff
@@ -0,0 +1,98 @@
+    const history = useHistory();
+    const invitations = useSelector((state: CombinedState) => state.invitations.invitations);
+    const [page, setPage] = useState(query.page);
+    const renderedInvitations = setUpInvitationsList(invitations, page);
```
Could you please explain? In reducer (GET_INVITATIONS_SUCCESS) I see it just assigns new invitations list, not append to existing one. So, it looks like only one page is stored in the storage.

bsekachev commented on cvat-ui/src/actions/invitations-actions.ts:
```diff
@@ -0,0 +1,124 @@
+                filter: `{"and":[{"==":[{"var":"user_id"},"${userID}"]}, {"==":[{"var":"accepted"},false]}]}`,
+            });
+            dispatch(invitationActions.getInvitationsSuccess(
+                invitations, invitations.length !== 0 && showNotification,
```
Shall we show notification if all the invitations are expired?

klakhov commented:
> ![image](https://private-user-images.githubusercontent.com/40690378/286863765-6cacf983-5d12-49b4-827b-7b3bd5247235.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzMzgxMjIsIm5iZiI6MTcwMTMzNzgyMiwicGF0aCI6Ii80MDY5MDM3OC8yODY4NjM3NjUtNmNhY2Y5ODMtNWQxMi00OWI0LTgyN2ItN2IzYmQ1MjQ3MjM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTMwVDA5NTAyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExOWYwMTFmOWE1OGVlYWNhM2RiZjY2NGVmZmYxMzIyNWU5ZDQ0ZmUxYjVlOGNjMzhlMmI0NDljMjQ1NjA0MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.L6JEhZIlzSg4BrEXZU0JZxn8FpuObNXj3rC_hbLKURk)
> 
> After I rejected invitation, inviting person clicks "Resend invitation" and gets 404 code. Maybe it is a rare case, but from the POV of user who resends invitation it may look like a problem.

I dont think thats really a case we should worry about. It can happen to any removable entity in cvat. If the task was deleted and some other user tries to open it without reloading the page, there also will be a 404 error.
![image](https://github.com/opencv/cvat/assets/50956430/ac3d28c4-5b02-49ca-af9d-f5ac66b1ccc3)

bsekachev commented:
@klakhov 

Just invitation from Docker. 

Let's use Accept/Decline (not Reject)

<img width="944" alt="image" src="https://github.com/opencv/cvat/assets/40690378/66c8a64f-6a73-4c87-8e6b-ee3446c0a1b2">

klakhov commented:
> @klakhov
> 
> Just invitation from Docker.
> 
> Let's use Accept/Decline (not Reject)
> 
> <img alt="image" width="944" src="https://private-user-images.githubusercontent.com/40690378/287218559-66c8a64f-6a73-4c87-8e6b-ee3446c0a1b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE0NDkyNjQsIm5iZiI6MTcwMTQ0ODk2NCwicGF0aCI6Ii80MDY5MDM3OC8yODcyMTg1NTktNjZjOGE2NGYtNmE3My00Yzg3LThlNmItZWUzNDQ2YzBhMWIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAxVDE2NDI0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiYTYzYTJmMDQ5MDY5ZWVmMzg3MDhiZjQwZDg2YjJmMDQ5MmIzNzQwMDNhNGMxYmIwMDc0YzJiZWFmYmYwZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.X-YXQDIpa1nZ4T4b76Ql25bHFvlJG1nPsQuT70nG5Tw">
> It did not require any additional confirmations when I clicked Accept
> 
> Also idea is that if Accept and Decline far away enough, we do not need confirmations on Decline also. It is hardly ever possible that user misclicks in this case. But maybe keep confirmation for Decline, this solution is up to you

Changed `reject` to `decline`. But lets keep confirmation for decline

bsekachev commented on cvat-ui/src/components/cvat-app.tsx:
```diff
@@ -300,7 +309,7 @@ class CVATApplication extends React.PureComponent<CVATAppProps & RouteComponentP
         }
 
-        if (user == null || !user.isVerified) {
+        if (user == null || !user.isVerified || !user.id) {
```
How can we have a user without id?

klakhov commented on cvat-ui/src/components/invitations-page/invitations-list.tsx:
```diff
@@ -0,0 +1,89 @@
+                description: 'Invalid page',
+            });
+        }
+    }, [invitations, query]);
```
![image](https://github.com/opencv/cvat/assets/50956430/78593625-7c87-41d7-ad33-e0a9624c8430)
Hm, yes there is already an error message. Removed that

bsekachev reviewed (approved):
Client part LGTM
Please, fix eslint issues:

/home/runner/work/cvat/cvat/cvat-ui/src/components/invitations-page/invitations-list.tsx
Error:    5:30  error  'useEffect' is defined but never used     @typescript-eslint/no-unused-vars
Error:   13:8   error  'notification' is defined but never used  @typescript-eslint/no-unused-vars

nmanovic commented on cvat/apps/organizations/serializers.py:
```diff
@@ -58,11 +61,12 @@ class InvitationReadSerializer(serializers.ModelSerializer):
     organization = serializers.PrimaryKeyRelatedField(
         queryset=Organization.objects.all(),
         source='membership.organization')
+    organization_info = BasicOrganizationSerializer(source='membership.organization')
```
@klakhov , let's meet and discuss some changes. I'm not sure why we need duplication of data here.

nmanovic replied:
Can we solve the problem with organization and organization_info in the PR?

nmanovic replied:
OK. Let's keep as is at the momemt

SpecLad commented on changelog.d/20231120_134843_klakhov_improve_invite_users.md:
```diff
@@ -0,0 +1,5 @@
+### Added
+
+- Implemented a feature that allows to invite users to organization using email
```
Didn't this feature already exist?

klakhov replied:
Acually you could invite only unregistered users using email and there were no UI for "internal" organization invitations for registered users.

Could you suggest better changelog description for the feature then?

klakhov replied:
Ok, let me explain a bit more.

Before this PR:
1. Organization staff invites user by entering his email. There are two cases: user already exists or doesnt exist. In first case invitation is auto accepted. In second one he receives email letter with registration link.
2. User follows the link and setups his account using registration form. Upon successful registration invitation to such organization is auto accepted.

After this PR (with enabled ORG_INVITATION_CONFIRM in settings):
1. Organization staff invites user by entering his email. There are two cases: user already exists or doesnt exist. In both cases email letter with login link is sent. 
2. User follows the link and logins/registers depending on if he had cvat account or not. Then he is redirected to page with list of his invitations where he can press a button to accept/reject the invitation.
3. Registered users doesn't really need to follow the link in the email, they can view and accept/reject their invitations using our UI. There is new section in Organization top menu called `invitations`. Accordingly, this is a page with a list of all invitations. User can check his invitations manually by visiting this page or by following a link in a popup notification when new invitation is received.

SpecLad replied:
> with enabled `ORG_INVITATION_CONFIRM` in settings

Say... how would one actually enable this setting? It seems to be hardcoded to "No".

klakhov replied:
By default it is disabled(like sending emails). I suppose we should enable it in our private production config alongside with email backend setup.

If someone wants to test it in development env, the setting must be changed manually.

SpecLad replied:
Okay, how about this:

> When the `ORG_INVITATION_CONFIRM` setting is enabled, organization invitations for existing users are no longer accepted automatically. Instead, the invitee can now review the invitation and accept or decline it.

klakhov replied:
Sounds good to me. With such wording `Changed` category suits more. Updated.

SpecLad commented on cvat/apps/iam/permissions.py:
```diff
@@ -95,10 +95,12 @@ def get_iam_context(request, obj):
-
     # FIXME: The primary app should know nothing about is_crowdsourcing plugin.
-    if organization and not request.user.is_superuser and membership is None and not iam_context.get('is_crowdsourcing', False):
-        raise PermissionDenied({'message': 'You should be an active member in the organization'})
+    # if organization and not request.user.is_superuser and membership is None and not iam_context.get('is_crowdsourcing', False):
+    #     # Allow inactive members view, accept, reject their invitations
+    #     view = request.parser_context.get('view')
+    #     if not (view.basename == 'invitation' and view.action in ['retrieve', 'accept', 'reject']):
+    #         raise PermissionDenied({'message': 'You should be an active member in the organization'})
```
Seems like this should be deleted?

klakhov replied:
Removed

SpecLad commented on cvat/schema.yml:
```diff
@@ -7357,8 +7362,14 @@ components:
           type: integer
+        expired:
+          type: boolean
+          readOnly: true
```
Shouldn't this also be `nullable`?

klakhov replied:
Yes, it can be null. Added.

SpecLad commented on cvat/apps/iam/serializers.py:
```diff
@@ -27,6 +32,44 @@ def get_cleaned_data(self):
+
+    def save(self, request):
+        adapter = get_adapter()
+        user = adapter.new_user(request)
```
Wouldn't it make more sense to do this after you verified that no dummy user exists?

klakhov replied:
Yes, in that case we dont need to call new_user. Fixed

SpecLad commented:
> Changed `reject` to `decline`. 

I can still see `reject` everywhere in the backend code. It should be updated too.

klakhov commented on cvat/apps/iam/rules/organizations.rego:
```diff
@@ -71,15 +71,15 @@ allow {
 } else := qobject {
     user := input.auth.user
-    qobject := [ {"owner_id": user.id}, {"members__user_id": user.id}, "|" ]
+    qobject := [{"members__user_id": user.id}, {"members__is_active": true}, "&", {"owner_id": user.id}, "|" ]
```
No, with `{"members__user_id": user.id}, {"members__is_active": true}, "&"` we check that the user with provided `user.id` is actually an active member. That helps us to filter organizations in which user is invited, but invitation is still pending

SpecLad replied:
You're right, I misremembered the Django filtering semantics.

klakhov commented on cvat/apps/iam/permissions.py:
```diff
@@ -96,10 +96,6 @@ def get_iam_context(request, obj):
         iam_context.update(builder_func(request, organization, membership))
 
-    # FIXME: The primary app should know nothing about is_crowdsourcing plugin.
-    if organization and not request.user.is_superuser and membership is None and not iam_context.get('is_crowdsourcing', False):
```
This logic is an optimization - the permission denied error is thrown here before calling the OPA. The idea behind deleting this part of code that all permission logic should be handled via OPA calls and we do not need a bunch of `if`s checking different cases when users is not a member but still needs organization resources.

So this logic is handled in OPA rules where we already have checks like `organizations.is_member`, `utils.is_organization`.

SpecLad replied:
What about the `is_crowdsourcing` condition? I don't see that anywhere else, including the private repo.

klakhov replied:
The idea behind this condition is to allow random user access to the job that is in organization. The private repo app setups `is_crowdsourcing` to `True` so all this condition resolves to `False` and throwing an error is skipped.

So if we remove throwing an error at all, we do not affect private part

SpecLad replied:
Does this mean that the `is_crowdsourcing` value is useless (since it's not used anywhere else)?

Marishka17 replied:
@SpecLad, Yes, it does, you can see that the `is_crowdsourcing` value setting is removed in 204 PR

SpecLad replied:
Alright, works for me.

klakhov commented:
It looks we have a couple of unstable e2e tests right now:
`actions_users/registration_involved/case_28_review_pipeline_feature. js`
`actions_projects_models/registration_involved/base_actions_project_task_user.js`

They failed in this pr a couple of times(after rerun everything is ok)
And they are failing in other PRs.

I will merge this PR as its already approved, and invistigate test problems separately.
